### PR TITLE
Fixes part of #16: Implementation of AdminAuthActivity and AddProfileActivity

### PR DESCRIPTION
**Pull request by jamesxu0**
## Explanation

Fixes part of #16: Implementation of AdminAuthActivity and AddProfileActivity. It also implements a custom ProfileInputView to easily display error messages. Note the base branch is not develop but refers to PR #326.

## Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR is made from a branch that is up-to-date with "develop".
- [x] The PR's branch is based on "develop" and not on any other branch.
- [x] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.

### CODE REVIEW

File: app/src/main/AndroidManifest.xml [hunk 1]
```diff
@@ -15,6 +15,8 @@
     <activity android:name=".testing.ContinuePlayingFragmentTestActivity" />
     <activity android:name=".home.HomeActivity" />
     <activity android:name=".player.audio.testing.AudioFragmentTestActivity" />
+    <activity android:name=".profile.AdminAuthActivity" />
+    <activity android:name=".profile.AddProfileActivity" />
```
Comment (rt4914) [inline app/src/main/AndroidManifest.xml]: Nit: Try to keep activities in alphabetical order.

File: app/src/main/java/org/oppia/app/profile/AddProfileActivityPresenter.kt [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+const val GALLERY_INTENT_RESULT_CODE = 1
+
+/** The presenter for [AddProfileActivity]. */
+@ActivityScope
+class AddProfileActivityPresenter @Inject constructor(
+  private val activity: AppCompatActivity,
+  private val profileManagementController: ProfileManagementController,
+  private val viewModelProvider: ViewModelProvider<AddProfileViewModel>
+) {
+  lateinit var uploadImageView: ImageView
```
Comment (rt4914) [inline app/src/main/java/org/oppia/app/profile/AddProfileActivityPresenter.kt]: This can be `private`

File: app/src/main/java/org/oppia/app/profile/AddProfileActivityPresenter.kt [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+  private val addViewModel by lazy {
+    getAddProfileViewModel()
+  }
+  private var selectedImage: Uri? = null
+  var allowDownloadAccess = false
```
Comment (rt4914) [inline app/src/main/java/org/oppia/app/profile/AddProfileActivityPresenter.kt]: This can be `private`

File: app/src/main/java/org/oppia/app/profile/AddProfileActivityPresenter.kt [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+    }
+
+    binding.infoIcon.setOnClickListener {
+      showInfoDialog()
+    }
+
+    addTextChangeListeners(binding)
+
+    binding.uploadImageButton.setOnClickListener {
+      val galleryIntent = Intent(Intent.ACTION_PICK, MediaStore.Images.Media.EXTERNAL_CONTENT_URI)
```
Comment (rt4914) [inline app/src/main/java/org/oppia/app/profile/AddProfileActivityPresenter.kt:87]: Are we supporting on gallery images and not camera pictures?

File: app/src/main/java/org/oppia/app/profile/ProfileInputView.kt [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+import android.view.LayoutInflater
+import android.view.View
+import android.widget.EditText
+import android.widget.LinearLayout
+import android.widget.TextView
+import androidx.databinding.BindingAdapter
+import androidx.databinding.DataBindingUtil
+import org.oppia.app.R
+import org.oppia.app.databinding.ProfileInputViewBinding
+import org.oppia.app.model.Profile
```
Comment (rt4914) [inline app/src/main/java/org/oppia/app/profile/ProfileInputView.kt]: remove unused imports.

**Review by rt4914 — APPROVED** [review]
Really awesome implementation. Just nit changes and suggest using `option+cmd+L` in every file after PR.

File: app/src/main/java/org/oppia/app/profile/AddProfileActivity.kt [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+
+  override fun onSupportNavigateUp(): Boolean {
+    finish()
+    return false
+  }
+
+  override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
+    super.onActivityResult(requestCode, resultCode, data)
+    if (requestCode == GALLERY_INTENT_RESULT_CODE && resultCode == Activity.RESULT_OK) {
+      addProfileFragmentPresenter.handleOnActivityResult(data)
```
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/profile/AddProfileActivity.kt]: Maybe just pass the code and data to the presenter to keep all logic there.

File: app/src/main/java/org/oppia/app/profile/AddProfileActivityPresenter.kt [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+    getAddProfileViewModel()
+  }
+  private var selectedImage: Uri? = null
+  var allowDownloadAccess = false
+  var inputtedPin = false
+  var inputtedConfirmPin = false
+
+  @ExperimentalCoroutinesApi
+  fun handleOnCreate() {
+    activity.title = "Add Profile"
```
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/profile/AddProfileActivityPresenter.kt]: Let’s set this via the manifest and using a strings.xml string so it can be translated.

File: app/src/main/java/org/oppia/app/profile/AddProfileActivityPresenter.kt [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+
+/** The presenter for [AddProfileActivity]. */
+@ActivityScope
+class AddProfileActivityPresenter @Inject constructor(
+  private val activity: AppCompatActivity,
+  private val profileManagementController: ProfileManagementController,
+  private val viewModelProvider: ViewModelProvider<AddProfileViewModel>
+) {
+  lateinit var uploadImageView: ImageView
+  private val addViewModel by lazy {
```
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/profile/AddProfileActivityPresenter.kt]: This sounds like a method. Could we rephrase it to sound a bit more like a noun or noun phrase?

File: app/src/main/java/org/oppia/app/profile/AddProfileActivityPresenter.kt (hunk 3 above)
Comment (BenHenning) [reply app/src/main/java/org/oppia/app/profile/AddProfileActivityPresenter.kt]: Ditto below. Prefer minimizing visibility.

File: app/src/main/java/org/oppia/app/profile/AddProfileViewModel.kt [hunk 9]
```diff
@@ -0,0 +1,20 @@
+package org.oppia.app.profile
+
+import androidx.databinding.ObservableField
+import org.oppia.app.activity.ActivityScope
+import org.oppia.app.viewmodel.ObservableViewModel
+import javax.inject.Inject
+
+@ActivityScope
+class AddProfileViewModel @Inject constructor() : ObservableViewModel() {
```
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/profile/AddProfileViewModel.kt:10]: Please add KDoc here and for other new classes.

File: app/src/main/java/org/oppia/app/profile/AddProfileActivityPresenter.kt [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+
+    addTextChangeListeners(binding)
+
+    binding.uploadImageButton.setOnClickListener {
+      val galleryIntent = Intent(Intent.ACTION_PICK, MediaStore.Images.Media.EXTERNAL_CONTENT_URI)
+      activity.startActivityForResult(galleryIntent, GALLERY_INTENT_RESULT_CODE)
+    }
+    uploadImageView = binding.uploadImageButton
+
+    binding.createButton.setOnClickListener {
```
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/profile/AddProfileActivityPresenter.kt:91]: I suggest breaking this up into several constituent helper methods to aid with readability.

File: app/src/main/java/org/oppia/app/profile/AddProfileActivityPresenter.kt [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+          addViewModel.confirmPinErrorMsg.set("")
+          inputtedConfirmPin = confirmPin.isNotEmpty()
+          setValidPin()
+        }
+      }
+      override fun afterTextChanged(confirmPin: Editable?) {}
+      override fun beforeTextChanged(p0: CharSequence?, start: Int, count: Int, after: Int) {}
+    })
+
+    binding.inputName.addTextChangedListener(object: TextWatcher {
```
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/profile/AddProfileActivityPresenter.kt]: These text watchers seem really repetitive. Could we factor them into a common helper that takes a lambda that performs the actions we want? Then it might be a bit simpler here, e.g.:

```kotlin
addTextChangedListener(binding.inputName) {
  confirmPin?.let { ... }
}
```

File: app/src/main/java/org/oppia/app/profile/AdminAuthActivityPresenter.kt [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+  private val activity: AppCompatActivity,
+  private val viewModelProvider: ViewModelProvider<AdminAuthViewModel>
+) {
+  private val authViewModel by lazy {
+    getAdminAuthViewModel()
+  }
+
+  /** Binds ViewModel and sets up text and button listeners. */
+  fun handleOnCreate() {
+    activity.title = "Add Profile"
```
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/profile/AdminAuthActivityPresenter.kt]: Ditto here & for other activity titles. These should be translatable.

File: app/src/main/java/org/oppia/app/profile/ProfileActivityPresenter.kt [hunk 13]
```diff
@@ -1,14 +1,23 @@
...
 @ActivityScope
-class ProfileActivityPresenter @Inject constructor(private val activity: AppCompatActivity) {
+class ProfileActivityPresenter @Inject constructor(
+  private val activity: AppCompatActivity,
+  private val profileManagementController: ProfileManagementController
+) {
+  /** Adds [ProfileChooserFragment] to view. */
+  @ExperimentalCoroutinesApi
   fun handleOnCreate() {
+    // Ensures that an admin profile is present
```
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/profile/ProfileActivityPresenter.kt]: Can you add a TODO to remove this that corresponds to a new issue that tracks introducing the user onboarding flow? This relates to #54, but we really need to just solve the onboarding problem rather than hardcoding an admin account like this.

File: app/src/main/java/org/oppia/app/profile/ProfileInputView.kt [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+
+class ProfileInputView @JvmOverloads constructor(
```
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/profile/ProfileInputView.kt:18]: Please add KDoc here & for other new classes in this PR.

File: app/src/sharedTest/java/org/oppia/app/profile/AddProfileActivityTest.kt [hunk 15]
```diff
@@ -0,0 +1,234 @@
...
+    GlobalScope.launch(Dispatchers.Main) {
+      profileTestHelper.initializeProfiles()
+    }
+    ActivityScenario.launch(AddProfileActivity::class.java).use {
+      onView(allOf(withId(R.id.input), isDescendantOfA(withId(R.id.input_name)))).perform(typeText("Sean"))
+      onView(withId(R.id.create_button)).perform(scrollTo()).perform(click())
+      onView(allOf(withId(R.id.error_text), isDescendantOfA(withId(R.id.input_name)))).check(matches(withText(context.getString(R.string.add_profile_error_name_not_unique))))
+
+      onView(allOf(withId(R.id.input), isDescendantOfA(withId(R.id.input_name)))).perform(typeText(" "))
+      onView(allOf(withId(R.id.error_text), isDescendantOfA(withId(R.id.input_name)))).check(matches(withText("")))
```
Comment (BenHenning) [inline app/src/sharedTest/java/org/oppia/app/profile/AddProfileActivityTest.kt:120]: Is this testing that the error gets cleared? If so, please split into its own test.

Ditto elsewhere.

File: app/src/sharedTest/java/org/oppia/app/profile/AdminAuthActivityTest.kt [hunk 16]
```diff
@@ -0,0 +1,154 @@
...
+
+  @Test
+  fun testAdminAuthActivity_inputIncorrectPassword_checkError_inputAgain_checkErrorIsGone() {
+    ActivityScenario.launch<AdminAuthActivity>(AdminAuthActivity.createAdminAuthActivityIntent(context, "12345")).use {
+      onView(allOf(withId(R.id.input), isDescendantOfA(withId(R.id.input_pin)))).perform(typeText("123"))
+      onView(withId(R.id.submit_button)).perform(click())
+      onView(allOf(withId(R.id.error_text), isDescendantOfA(withId(R.id.input_pin)))).check(matches(withText(context.resources.getString(R.string.admin_auth_incorrect))))
+
+      onView(allOf(withId(R.id.input), isDescendantOfA(withId(R.id.input_pin)))).perform(typeText("4"))
+      onView(allOf(withId(R.id.error_text), isDescendantOfA(withId(R.id.input_pin)))).check(matches(withText("")))
```
Comment (BenHenning) [inline app/src/sharedTest/java/org/oppia/app/profile/AdminAuthActivityTest.kt:89]: Ditto here--ensuring the error is cleared out to be a separate test.

**Review by BenHenning — APPROVED** [review]
I skimmed through the Kotlin portions, but didn't look at the XML files in detail. Otherwise, I defer to @rt4914 for verifying the UI structural & correctness aspects of this PR.

Please let me know & reassign to me if you'd like me to look at anything specific.

File: app/src/main/java/org/oppia/app/profile/AddProfileActivity.kt (hunk 6 above)
Comment (jamesxu0) [reply app/src/main/java/org/oppia/app/profile/AddProfileActivity.kt]: Moved logic to presenter.

File: app/src/main/java/org/oppia/app/profile/AddProfileActivityPresenter.kt (hunk 2 above)
Comment (jamesxu0) [reply app/src/main/java/org/oppia/app/profile/AddProfileActivityPresenter.kt]: Updated to private.

File: app/src/main/java/org/oppia/app/profile/AddProfileActivityPresenter.kt (hunk 8 above)
Comment (jamesxu0) [reply app/src/main/java/org/oppia/app/profile/AddProfileActivityPresenter.kt]: Updated to profileViewModel.

File: app/src/main/java/org/oppia/app/profile/AddProfileActivityPresenter.kt (hunk 3 above)
Comment (jamesxu0) [reply app/src/main/java/org/oppia/app/profile/AddProfileActivityPresenter.kt]: Updated to private.

File: app/src/main/java/org/oppia/app/profile/AddProfileActivityPresenter.kt (hunk 7 above)
Comment (jamesxu0) [reply app/src/main/java/org/oppia/app/profile/AddProfileActivityPresenter.kt]: Updated to use a string resource.

File: app/src/main/java/org/oppia/app/profile/AddProfileActivityPresenter.kt (hunk 10 above)
Comment (jamesxu0) [reply app/src/main/java/org/oppia/app/profile/AddProfileActivityPresenter.kt:91]: Added helper functions.

File: app/src/main/java/org/oppia/app/profile/AddProfileActivityPresenter.kt (hunk 4 above)
Comment (jamesxu0) [reply app/src/main/java/org/oppia/app/profile/AddProfileActivityPresenter.kt:87]: Not currently. I will add support in a future PR.

File: app/src/main/java/org/oppia/app/profile/AddProfileActivityPresenter.kt (hunk 11 above)
Comment (jamesxu0) [reply app/src/main/java/org/oppia/app/profile/AddProfileActivityPresenter.kt]: Thanks for the suggestion! This made the code a lot easier to read.

File: app/src/main/java/org/oppia/app/profile/AddProfileViewModel.kt (hunk 9 above)
Comment (jamesxu0) [reply app/src/main/java/org/oppia/app/profile/AddProfileViewModel.kt:10]: Added KDocs.

File: app/src/main/java/org/oppia/app/profile/AdminAuthActivityPresenter.kt (hunk 12 above)
Comment (jamesxu0) [reply app/src/main/java/org/oppia/app/profile/AdminAuthActivityPresenter.kt]: Updated to use a string resource.

File: app/src/main/java/org/oppia/app/profile/ProfileActivityPresenter.kt (hunk 13 above)
Comment (jamesxu0) [reply app/src/main/java/org/oppia/app/profile/ProfileActivityPresenter.kt]: Added TODO for introducing the user on boarding flow. https://github.com/oppia/oppia-android/issues/482

File: app/src/main/java/org/oppia/app/profile/ProfileInputView.kt (hunk 5 above)
Comment (jamesxu0) [reply app/src/main/java/org/oppia/app/profile/ProfileInputView.kt]: Removed.

File: app/src/main/java/org/oppia/app/profile/ProfileInputView.kt (hunk 14 above)
Comment (jamesxu0) [reply app/src/main/java/org/oppia/app/profile/ProfileInputView.kt:18]: Added KDoc.

File: app/src/sharedTest/java/org/oppia/app/profile/AddProfileActivityTest.kt (hunk 15 above)
Comment (jamesxu0) [reply app/src/sharedTest/java/org/oppia/app/profile/AddProfileActivityTest.kt:120]: Updated to separate tests.

File: app/src/sharedTest/java/org/oppia/app/profile/AdminAuthActivityTest.kt (hunk 16 above)
Comment (jamesxu0) [reply app/src/sharedTest/java/org/oppia/app/profile/AdminAuthActivityTest.kt:89]: Updated to separate test.

File: app/src/main/AndroidManifest.xml (hunk 1 above)
Comment (jamesxu0) [reply app/src/main/AndroidManifest.xml]: Updated.
